### DISCUSSION TITLE
Probably a stupid question about service activation (i'm such a code-hater)

### DISCUSSION THREAD

**Discussion by MHKnowles** [BottlecapDave/HomeAssistant-OctopusEnergy#633] [General]
Hi all,

I see that there's a service to join a new octoplus saving session, and there's some example code.  However, I can't see how in the automation visual editor you'd call the service as a trigger?  and also if I change to the YAML editor and paste in the example code I get a vague error relating to the code syntax.

Can anyone tell me (preferably in the visual editor cos I can use that to an extent) how I should go about setting up an automation to join a new saving session?

Second (bonus) part of the question - can anyone tell me how to create an automation to trigger a charge event on my batteries x minutes before the event start time, where x is the length of the saving event in minutes? (so that I can trigger my batteries to charge from the grid before the event)

Thanks in advance you lovely people!

**Comment by RobinXe** [comment]
Hi!

In answer to the first part of your question, I had the same confusion. It wasn't until I pored over the code that I realised the 'octopus_energy_new_octoplus_saving_session' event is aliased in const.py to 'EVENT_NEW_SAVING_SESSION'. If you trigger on a 'octopus_energy_new_octoplus_saving_session' event, it will work, you'll just have to type the event name (accurately) into the visual editor manual event.

I'm also looking at the second part of your query, but with the intention of putting _just_ enough into the battery, if there is too little. How much is optimal depends on which tariff you are on, and overcharging could be sub-optimal. As an example, if you're on agile/flux, and the saving session is at 1700/1800, then charging just before would involve importing with a peak surcharge, you'd be better off 'overcharging' earlier, and running from battery til the saving session. Conversely, if you're on Go/IO, then you don't want to import any more than the minimum to ensure maximal export during the session, as any extra will be subject to the losses of going through your battery, so you'd be better off importing later in the day, and using it directly.

If you just want to start charging based on time before the next saving session, then you'll probably need a helper based off the next saving session datetime minus a timedelta. I don't think you'll achieve that without code. I'm keen to be proven wrong though, and am interested in other solutions!

**Reply by MHKnowles** [reply]
Thanks - I solved the first part of my query a very short while ago by educating myself about Blueprints and how they work in HA.  I managed to import the blueprint from the blueprint page here and it seems to work (at least it didn't give me any errors).

As regards just enough energy, I charge my batteries to 100% on the IO tariff since my export rate is higher than the off-peak import so even if it's a sunny day and the battery stays charged i'm making rather than losing money on having the battery charged to full. It doesn't necessarily last the day anyway given my usage, so i'm normally down to about 20% by the time the saving sessions have typically kicked in.  I have services for charging and discharging the battery meaning that I can charge at a set kW level to a set state of charge and vice versa for discharge, so assuming I charge at 5000kW for 90 minutes prior to a 90 minute saving session then I can discharge at the same rate during the session...that's my logic to it anyway.  So yes, you're right i'll probably need a snippet of code, I just don't know what code :)

**Reply by RobinXe** [reply]
In that case you're in a similar situation to me (I'm on Go until i get my Zappi installed in Jan). I'm working on something to charge (when necessary) prior to a session. I'll loop you in once i have it working.

What battery/inverter do you have?

**Reply by MHKnowles** [reply]
I have a Huawei Sun2000 inverter with 14 x 370W JA Solar panels, and a Huawei Luna2000 battery at 10kW.
I used to be on Go until I realised that my Tesla was compatible with IO and gave me a larger night-time backup window at a lower cost, so win-win.  I don't bother with the intelligent charging because I have a dumb charger which can only work with schedules (don't buy a Qubev charger if you want to control it - you need to run an ethernet cable to the device despite the rest of it working over wifi!) and with the wife having an i3 I didn't want her coming home, plugging in, and charging on peak usage.

**Reply by RobinXe** [reply]
Ah, that could potentially add a layer of complexity, as I'm using a SolarEdge system. If you have an integration to control the battery through Home Assistant though, we should be able to generalise it. I'll let you know when I have something working.

**Reply by MHKnowles** [reply]
As long as I can get a trigger that can trigger an event x minutes before the start then I should be golden with that.  I don't expect it all to be perfect, because currently there's a fair amount of headroom on the Octoplus events in terms of reward so I'm prepared to play around once I'm in the right ballpark.

**Reply by RobinXe** [reply]
There's ways to do that. Take a look at templates in HA, and timedelta, and you should be able to knock something up. I'll share what I come up with when I come up with it.
